Support quantized int8 and uint8 in TFLu mean operator

patriklaurell opened the pull request.

patriklaurell committed:
Support quantized int8 and uint8 in TFLu mean operator

hajuho commented on tensorflow/lite/micro/kernels/reduce.cc:
```diff
@@ -30,6 +31,16 @@ namespace reduce {
+};
+
+void* InitMean(TfLiteContext* context, const char* buffer, size_t length) {
+  auto* op_data = new OpData();
```
void* raw;
  context->AllocatePersistentBuffer(context, sizeof(OpData), &raw);
  return raw;

advaitjain replied:
yes, please make this change. We can not have dynamic memory allocation for micro and the AllocatePersistentBuffer API is the way to go.

patriklaurell replied:
fixed in latest commit

hajuho commented on tensorflow/lite/micro/kernels/reduce.cc:
```diff
@@ -30,6 +31,16 @@ namespace reduce {
+};
+
+void* InitMean(TfLiteContext* context, const char* buffer, size_t length) {
+  auto* op_data = new OpData();
```
void* raw;
  context->AllocatePersistentBuffer(context, sizeof(OpData), &raw);
  return raw;

hajuho commented on tensorflow/lite/micro/kernels/reduce.cc:
```diff
@@ -109,18 +130,76 @@ TfLiteStatus EvalMean(TfLiteContext* context, TfLiteNode* node) {
+                input->dims->size, GetTensorData<int8_t>(output),
+                output->dims->data, output->dims->size,
+                GetTensorData<int>(axis), num_axis, params->keep_dims,
+                temp_index, resolved_axis, GetTensorData<int>(output)));
```
GetTensorData<int8_t>(output)

patriklaurell replied:
The last argument is for the temporary sum that is then divided by the number of elements to make up the mean. Does this not need to be of type with more than 8 bits to handle the potential overflow of the sum? @hajuho

hajuho replied:
I see. I missed it.
But, does the "output" tensor have enough capacity to store the array of int values? Temporary buffer is used for TFLite https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/reduce.cc#L428, but looks like GetTemporary is not available in micro https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/kernels/svdf.cc#L465.
@advaijain Do you know how to handle it?

hajuho replied:
Sorry, I made a typo. @advaitjain Do you have any idea on this?

patriklaurell replied:
Any update on this? @hajuho @advaitjain

patriklaurell replied:
In latest commit I use context->RequestScratchBufferInArena to allocate this tensor

hajuho commented on tensorflow/lite/micro/kernels/reduce.cc:
```diff
@@ -109,18 +130,76 @@ TfLiteStatus EvalMean(TfLiteContext* context, TfLiteNode* node) {
+                GetTensorData<int8_t>(output), output->params.zero_point,
+                output->params.scale, output->dims->data, output->dims->size,
+                GetTensorData<int>(axis), num_axis, params->keep_dims,
+                temp_index, resolved_axis, GetTensorData<int>(output), false));
```
ditto

patriklaurell replied:
same as above

hajuho commented on tensorflow/lite/micro/kernels/reduce.cc:
```diff
@@ -109,18 +130,76 @@ TfLiteStatus EvalMean(TfLiteContext* context, TfLiteNode* node) {
+                input->dims->size, GetTensorData<uint8_t>(output),
+                output->dims->data, output->dims->size,
+                GetTensorData<int>(axis), num_axis, params->keep_dims,
+                temp_index, resolved_axis, GetTensorData<int>(output)));
```
GetTensorData<uint8_t>(output)

patriklaurell replied:
same as above

hajuho commented on tensorflow/lite/micro/kernels/reduce.cc:
```diff
@@ -109,18 +130,76 @@ TfLiteStatus EvalMean(TfLiteContext* context, TfLiteNode* node) {
+                GetTensorData<uint8_t>(output), output->params.zero_point,
+                output->params.scale, output->dims->data, output->dims->size,
+                GetTensorData<int>(axis), num_axis, params->keep_dims,
+                temp_index, resolved_axis, GetTensorData<int>(output), false));
```
ditto

patriklaurell replied:
same as above

hajuho commented:
Meghna,
could you review this pull request?

MeghnaNatraj commented:
@advaitjain could you take a look at this?

advaitjain commented on tensorflow/lite/kernels/internal/reference/reduce.h:
```diff
@@ -354,11 +354,11 @@ inline bool QuantizedMeanOrSum(const T* input_data, int32 input_zero_point,
 
   // Calculate mean by dividing output_data by num of aggregated element.
-  U num_elements_in_axis = 1;
+  size_t num_elements_in_axis = 1;
```
is there a particular error that the template param was causing? If not, we should keep the template param.

patriklaurell replied:
Got a compilation error for comparing signed and unsigned. Removed this change and fixed it in function call

hajuho replied:
Still got compilation errors for comparing signed int and unsigned int with google internal build toolchain.
Asked to the original authors about this "U num_elements_in_axis = 1;" line and they agreed size_t is better for this variable.
Could you add this change again?

patriklaurell replied:
Done. Added this change again in latest commit.

advaitjain reviewed (changes requested):
Thanks for the PR. Please also address the comments from hajuho@

gbaned commented:
@patriklaurell Can you please check @advaitjain comments and keep us posted. Thanks!

patriklaurell committed:
Addressed review comments

patriklaurell commented:
Rebased on tip of master to get latest changes to the AllocatePersistentBuffer API, to get the unit test passing.

hajuho commented on tensorflow/lite/micro/kernels/reduce.cc:
```diff
@@ -61,12 +63,19 @@ TfLiteStatus PrepareSimple(TfLiteContext* context, TfLiteNode* node) {
+    QuantizeMultiplier(real_multiplier, &op_data->multiplier, &op_data->shift);
+  }
+
+  TfLiteTensor* output = GetOutput(context, node, 0);
```
Line 68 can be reused.

patriklaurell replied:
Fixed

patriklaurell committed:
Addressed hajuho's comment

hajuho commented on tensorflow/lite/micro/kernels/reduce.cc:
```diff
@@ -18,6 +18,7 @@ limitations under the License.
 #include "tensorflow/lite/c/builtin_op_data.h"
 #include "tensorflow/lite/c/common.h"
 #include "tensorflow/lite/kernels/internal/quantization_util.h"
+#include "tensorflow/lite/kernels/internal/reference/integer_ops/mean.h"
```
Need to add  this header file to https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/tools/make/Makefile#L176

patriklaurell replied:
Fixed

hajuho reviewed (commented):
Fails with segmentation fault for
make -f tensorflow/lite/micro/tools/make/Makefile test
and
tensorflow/tools/ci_build/ci_build.sh MICRO "tensorflow/lite/micro/tools/ci_build/test_all.sh"

Please test and fix it.

patriklaurell committed:
Updated Makefile

hajuho commented:
Still fails test
make -f tensorflow/lite/micro/tools/make/Makefile test_kernel_reduce_test

advaitjain@ could you review this again?

patriklaurell commented:
@hajuho running 
```
make -f tensorflow/lite/micro/tools/make/Makefile clean
make -f tensorflow/lite/micro/tools/make/Makefile test_kernel_reduce_test
```
I got the test to pass with the following output.
```
tensorflow/lite/micro/testing/test_linux_binary.sh tensorflow/lite/micro/tools/make/gen/linux_x86_64/bin/kernel_reduce_test '~~~ALL TESTS PASSED~~~'
tensorflow/lite/micro/tools/make/gen/linux_x86_64/bin/kernel_reduce_test: PASS
```

patriklaurell committed:
Re-added changes to internal/reference/reduce.h

gbaned commented:
@patriklaurell Can you please resolve conflicts? Thanks!

hajuho commented:
@patriklaurell 
The test is successful with the baseline code of this PR, but fails when I rebased to the HEAD.
Could you take a look?

hajuho commented on tensorflow/lite/micro/kernels/reduce_test.cc:
```diff
@@ -128,6 +128,48 @@ void TestMeanFloatInput4D(const int* input_dims_data, const float* input_data,
+  // Initialize tensors
+  constexpr int tensors_size = 3;
+  TfLiteTensor tensors[] = {
+      CreateQuantizedTensor(input_data, input_data_quant, input_dims,
```
The signature of this function has changed.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/testing/test_utils.h#L88

patriklaurell replied:
fixed

hajuho commented on tensorflow/lite/micro/kernels/reduce_test.cc:
```diff
@@ -128,6 +128,48 @@ void TestMeanFloatInput4D(const int* input_dims_data, const float* input_data,
+  const int input_dims_count = ElementCount(*input_dims);
+
+  // Allocate arrays for quantized tensors
+  T output_data_quant[output_dims_count];
```
This is blocked in Google internal toolchain.
error: variable length array used [-Werror,-Wvla]

Please fix other lines too.

patriklaurell replied:
fixed

hajuho commented on tensorflow/lite/kernels/internal/reference/reduce.h:
```diff
@@ -354,7 +354,7 @@ inline bool QuantizedMeanOrSum(const T* input_data, int32 input_zero_point,
 
   // Calculate mean by dividing output_data by num of aggregated element.
-  U num_elements_in_axis = 1;
+  size_t num_elements_in_axis = 1;
```
Thank you.
Could you change the line 361 std::numeric_limits<size_t>::max()?
Also the line 187 and 191?

patriklaurell replied:
fixed

patriklaurell committed:
Address review comments

patriklaurell commented:
> @patriklaurell
> The test is successful with the baseline code of this PR, but fails when I rebased to the HEAD.
> Could you take a look?

@hajuho I rebased on master to get latest changes to CreateQuantizedTensor and addressed your comments. All tests pass for me now standing on commit 48dff25

rthadur commented:
@hajuho can you please help merge this PR ?

hajuho commented:
For now, this PR fails with next command which is a part of google internal tests.
tensorflow/tools/ci_build/ci_build.sh MICRO "tensorflow/lite/micro/tools/ci_build/test_all.sh"

/workspace/tensorflow/lite/micro/tools/make/downloads/gcc_embedded/bin/../lib/gcc/arm-none-eabi/7.3.1/thumb/v7e-m/libgcc.a(unwind-arm.o): In function `get_eit_entry':
unwind-arm.c:(.text+0x138): undefined reference to `__exidx_end'
unwind-arm.c:(.text+0x13c): undefined reference to `__exidx_start'
collect2: error: ld returned 1 exit status
tensorflow/lite/micro/tools/make/Makefile:358: recipe for target 'tensorflow/lite/micro/tools/make/gen/stm32f4_cortex-m4/bin/kernel_reduce_test' failed
make: *** [tensorflow/lite/micro/tools/make/gen/stm32f4_cortex-m4/bin/kernel_reduce_test] Error 1
make: *** Waiting for unfinished jobs....

@patriklaurell Can you also take a look into this?

hajuho commented on tensorflow/lite/micro/kernels/reduce_test.cc:
```diff
@@ -126,6 +126,48 @@ void TestMeanFloatInput4D(const int* input_dims_data, const float* input_data,
+  const int input_dims_count = ElementCount(*input_dims);
+
+  // Allocate arrays for quantized tensors
+  T* output_data_quant = new T[output_dims_count];
```
Looks like these dynamic allocation has problem in some targets.
If I change this to "T output_data_quant[100];", the error disappears.
However, fixing the array length to some numeric value might not be a correct way to handle this.
@advaitjain Do you know why this happens? And is there any proper way to allocate dynamically in tf micro targets?

patriklaurell replied:
Any update on this? @hajuho @advaitjain

hajuho replied:
@petewarden Can you give a guide for allocating temporary buffers like this?

petewarden replied:
Sorry for the delay. I'd actually recommend you pass in the output data array as an argument into the function, since the calling code will know the size needed as a constant. It's a bit less convenient, but not all of our platforms support memory allocation, so we can't use new/malloc/etc.

gbaned commented:
@patriklaurell Can you please check @petewarden's comments and keep us posted ? Thanks!

gbaned commented:
@patriklaurell, Any update on this PR? Please. Thanks!

gbaned commented:
@patriklaurell, Any update on this PR? Please. Thanks!

patriklaurell committed:
Merge branch 'master' into tflu-8bit-mean

patriklaurell committed:
Updated implementation for compatibility with TfLiteEvalTensor

patriklaurell committed:
Addressed petewarden's comment

patriklaurell commented:
Sorry for the delay. I've been on my summer vacation the last weeks. I now resolved the conflicts with master and addressed @petewarden's comment.

rthadur commented on tensorflow/lite/micro/kernels/reduce.cc:
```diff
@@ -18,6 +18,7 @@ limitations under the License.
 #include "tensorflow/lite/c/builtin_op_data.h"
 #include "tensorflow/lite/c/common.h"
 #include "tensorflow/lite/kernels/internal/quantization_util.h"
+#include "tensorflow/lite/kernels/internal/reference/integer_ops/mean.h"
```
`tensorflow/lite/micro/kernels/reduce.cc:21:10: fatal error: '/tensorflow/lite/kernels/internal/reference/integer_ops/mean.h' file not found
#include "tensorflow/lite/kernels/internal/reference/integer_ops/mean.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.`

patriklaurell replied:
What command is run to reproduce this error?

rthadur replied:
@patrikerdes here is the command https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/benchmarks/BUILD#L21

rthadur replied:
This is still failing internally , can you please check

patriklaurell replied:
Running `bazel build //tensorflow/lite/micro/benchmarks:conv_benchmark` locally I do not encounter the problem. Could you provide more detailed instructions for how to reproduce this error? @rthadur

patriklaurell committed:
Fixed errors in tests

patriklaurell committed:
Enable tests to pass when running through make

gbaned commented:
@patriklaurell  Can you please check @advaitjain's comments and keep us posted ? Thanks!

advaitjain committed:
Do not exclde integer_ops/mean.h from tf_lite_static_memory bazel builds.

advaitjain commented:
I pushed a commit to @patriklaurell's branch to fix the error due to an internal build variant that is not currently exposed externally and was causing head scratching [here](https://github.com/tensorflow/tensorflow/pull/38634#discussion_r474535671)

Here are my commands, adapted from [this gist](https://gist.github.com/elfrank/c08256de9c15e41e1781):
```
git remote add patriklaurell git@github.com:patriklaurell/tensorflow.git
git fetch patriklaurell
git checkout -b patriklaurell-tflu-8bit-mean patriklaurell/tflu-8bit-mean

# make the fixes and commit to the local branch

git push patriklaurell tflu-8bit-mean
```
